src/libtego/source/utils/CryptoKey.cpp

```diff
@@ -174,19 +174,36 @@ QByteArray torControlHashedPassword(const QByteArray &password)
 
     int count = (16u + (96 & 15)) << ((96 >> 4) + 6);
 
-    SHA_CTX hash;
-    SHA1_Init(&hash);
+    EVP_MD_CTX *mdctx;
+    mdctx = EVP_MD_CTX_new();
```

Comment:
Can you fix the following code such that in the event of throw'ing mdctx does not leak?

Reply:
ideally by putting mdctx in a unique_ptr<EVP_MD_CTX> . You'll need to construct it such that EVP_MD_CTX_free is the provided cleanup method.

It should be something like:

```c++
EVP_MD_CTX* pctx = nullptr;
pctx = EVP_MD_CTX_new();
unique_ptr<EVP_MD_CTX, decltype(EVP_MD_CTX_free)> ctx(pctx, EVP_MD_CTX_free);
```

See https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr

Reply:
like 2852d37768a1b882c8e56353b1b2b64ae2b9bfe8?

i have no experience here, so i have no idea if this works.
if it looks good to you, i would squash the commits

Reply:
Sorry not quite.

I misread how EVP_MD_CTX_new() worked (thought it was a C-style out param type situation).

We can get rid of the intermediate `mdctx_hidden` variable entirely and just pass the output of `EVP_MD_CTX_new()` direclty into the ctx `unique_ptr`. Then in subsequent calls to various openssl methods, we need to call `ctx.get()` since iirc `unique_pr` does not auto downcast to the owned pointer type.

Reply:
d684f07de03dcc9b96ccddcf69456dc0ff0d0a30?

i hope the `#define mdctx mdctx_ptr.get()` works, it gives a minimal diff

Reply:
06202e1bb8268e745425b8f92726ccdefbd10adb?